File: src/main/java/com/nobot/plugin/dice/expressionAnalyzer/RandomExpression.java
package com.nobot.plugin.dice.expressionAnalyzer;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.var;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.Random;

/**
 * 这是最小的随机数表达式单位
 */
public class RandomExpression implements SpecialSymbol, SingleExpression
{
	private Random random;
	private int diceNum=1,diceDownLimit,diceUpperLimit;
	private int bonusNum,punishNum;
	private int bigDiceNum,smallDiceNum;

	private boolean onlyNumber;
	@Getter(AccessLevel.PACKAGE)
	private int bonusOrPunishMode, maxOrMinMode;
	private int tureBPNum,tureMaxOrMinNum;

	private int[] randomDiceArray,extraDiceArray,afterDiceArray,selectDiceArray;
	@Getter
	private int result;

	private String resource;
	private ArrayList<Object> list;

	//FIXME:将来将这里和下一个方法结合在一起，现在混淆了用法
	protected RandomExpression(Random random,String resource)
	{
		this.random=random;
		this.resource=resource;
		list=new ArrayList<>();
		if(resource.contains(String.valueOf(symbol_d)))
		{
			onlyNumber=false;
			checkExpression();
			randomDiceArray = new int[diceNum];
			if (bonusOrPunishMode != NO_PUNISH_AND_BONUS)
			{
				extraDiceArray = new int[tureBPNum];
				afterDiceArray = new int[randomDiceArray.length];
			}
			if (maxOrMinMode != GET_ALL_DICE_MODE)
				selectDiceArray = new int[tureMaxOrMinNum];
		}
		else
		{
			onlyNumber = true;
			try
			{
				result=Integer.parseInt(resource);
			}
			catch (NumberFormatException e)
			{
				throw new ExpressionException(resource,"不存在D时无法使用其他修饰符");
			}
		}
	}

	private void checkExpression()
	{
		var chars=resource.toCharArray();
		var sb=new StringBuilder();
//		分离数字和字符
		for (var c:chars)
		{
			switch (c)
			{
				case '0':
				case '1':
				case '2':
				case '3':
				case '4':
				case '5':
				case '6':
				case '7':
				case '8':
				case '9':
					sb.append(c);
					break;
				default:
					if(sb.length()!=0)
					{
						list.add(Integer.parseInt(sb.toString()));
						sb = new StringBuilder();
					}
					list.add(c);
					break;
			}
		}
		if(sb.length()!=0)
			list.add(Integer.parseInt(sb.toString()));
//		识别并处理控制符
		for (int i=0;i<list.size();i++)
		{
			var object=list.get(i);
			if(object instanceof Character)
			{
				switch ((char)object)
				{
					case symbol_b:
						var bonus=getInteger(i+1);
						if(bonus==null)
							bonusNum=1;
						else
							bonusNum=bonus;
						break;
					case symbol_p:
						var punish=getInteger(i+1);
						if(punish==null)
							punishNum=1;
						else
							punishNum=punish;
						break;
					case symbol_k:
						var bigDice=getInteger(i+1);
						if(bigDice==null)
							bigDiceNum=1;
						else
							bigDiceNum=bigDice;
						break;
					case symbol_q:
						var smallDice=getInteger(i+1);
						if(smallDice==null)
							smallDiceNum=1;
						else
							smallDiceNum=smallDice;
						break;
					case symbol_d:
						var randomTime=getInteger(i-1);
						if(randomTime==null)
							diceNum=1;
						else
							diceNum=randomTime;
						var randomSize=getInteger(i+1);
						if(randomSize==null)
							diceUpperLimit=100;
						else
							diceUpperLimit=randomSize;
						diceDownLimit=1;
						break;
					case symbol_t:
						var downLimit=getInteger(i-1);
						if(downLimit==null)
							diceDownLimit=1;
						else
							diceDownLimit=downLimit;
						var upLimit=getInteger(i+1);
						if(upLimit==null)
							diceUpperLimit=100;
						else
							diceUpperLimit=upLimit;
						break;
					default:
						throw new ExpressionException(resource,"无法识别的控制符号："+ (char) object);
				}
			}
		}
//		获得正式的奖励或者惩罚数量，两者相互抵消，检测是否是百分骰
		tureBPNum=bonusNum-punishNum;
		if(tureBPNum>0)
			bonusOrPunishMode =BONUS_MODE;
		else if(tureBPNum<0)
		{
			bonusOrPunishMode = PUNISH_MODE;
			tureBPNum=-tureBPNum;
		}
		else
			bonusOrPunishMode =NO_PUNISH_AND_BONUS;
		if((diceDownLimit!=1||diceUpperLimit!=100||diceNum!=1)&&bonusOrPunishMode!=NO_PUNISH_AND_BONUS)
			throw new ExpressionException(resource,"非单独一个百分骰不启用b和p控制符号");
//		获得正式的取大或者取小值，同时存在报错，检测是否取值超过了总值
		if(bigDiceNum!=0&&smallDiceNum!=0)
			throw new ExpressionException(resource,"同时存在k和q控制符");
		if(bigDiceNum>0)
		{
			maxOrMinMode=GET_MAX_DICE_MODE;
			tureMaxOrMinNum=bigDiceNum;
		}
		else if(smallDiceNum>0)
		{
			maxOrMinMode=GET_MIN_DICE_MODE;
			tureMaxOrMinNum=smallDiceNum;
		}
		else
			maxOrMinMode=GET_ALL_DICE_MODE;
		if(tureMaxOrMinNum>diceNum)
			throw new ExpressionException(resource,"取最"+(maxOrMinMode==GET_MAX_DICE_MODE?'大':'小')+"骰子数目超出总数目");
	}

	@Override
	public void calculation()
	{
		if(onlyNumber)
			return;
		for (int i=0;i<diceNum;i++)
		{
			if(diceUpperLimit==diceDownLimit)
			{
				randomDiceArray[i] = diceUpperLimit;
				continue;
			}
			randomDiceArray[i]=random.nextInt(diceUpperLimit-diceDownLimit+1)+diceDownLimit;
		}

		if (bonusOrPunishMode!=NO_PUNISH_AND_BONUS)
		{
			for (int i=0;i<tureBPNum;i++)
				extraDiceArray[i]=random.nextInt(10);
			int replaceNum=bonusOrPunishMode==BONUS_MODE?10:-1;
			for(var i:extraDiceArray)
			{
				if(bonusOrPunishMode==BONUS_MODE)
				{
					if (i<replaceNum)
						replaceNum=i;
				}
				else
				{
					if (i > replaceNum)
						replaceNum = i;
				}
			}
			for (int i = 0; i < randomDiceArray.length; i++)
			{
				int num=randomDiceArray[i];
				int tenDigit,singleDigit;
				tenDigit=num/10;
				singleDigit=num%10;
				if(bonusOrPunishMode==BONUS_MODE)
				{
					if (tenDigit>replaceNum)
					{
						tenDigit = replaceNum;
						afterDiceArray[i]=tenDigit*10+singleDigit;
					}
				}
				else if(bonusOrPunishMode==PUNISH_MODE)
				{
					if(tenDigit<replaceNum)
					{
						tenDigit = replaceNum;
						afterDiceArray[i]=tenDigit*10+singleDigit;
					}
				}
				else
					afterDiceArray[i]=num;
			}
		}

		if(maxOrMinMode!=GET_ALL_DICE_MODE)
		{
			int[] array;
			if(bonusOrPunishMode!=NO_PUNISH_AND_BONUS)
				array=afterDiceArray;
			else
				array=randomDiceArray;
			if(maxOrMinMode==GET_MIN_DICE_MODE)
				selectDiceArray=Arrays.stream(array).sorted().limit(selectDiceArray.length).toArray();
			else
				selectDiceArray=Arrays.stream(array).boxed().sorted(Comparator.reverseOrder())
						.limit(selectDiceArray.length).mapToInt(value -> value).toArray();
			result= Arrays.stream(selectDiceArray).sum();
		}
		if(bonusOrPunishMode==NO_PUNISH_AND_BONUS)
			result= Arrays.stream(randomDiceArray).sum();
		else
			result= Arrays.stream(afterDiceArray).sum();
	}

	@Override
	public String getShowExpression()
	{
		if(onlyNumber)
			return String.valueOf(result);
		var builder=new StringBuilder();
		if(randomDiceArray.length!=1)
		{
			builder.append('(');
			for (int i : randomDiceArray)
				builder.append(i).append("+");
			builder.delete(builder.length()-1,builder.length()).append(')');
		}
		else
			builder.append(randomDiceArray[0]);
		if(bonusOrPunishMode!=NO_PUNISH_AND_BONUS)
		{
			builder.append('{');
			if(bonusOrPunishMode==BONUS_MODE)
				builder.append(symbol_b);
			else
				builder.append(symbol_p);
			builder.append(':');
			for (int i : extraDiceArray)
				builder.append(i).append(',');
			builder.delete(builder.length()-1,builder.length()).append('}');
		}
		if(maxOrMinMode!=GET_ALL_DICE_MODE)
		{
			builder.append('[');
			if(maxOrMinMode==GET_MAX_DICE_MODE)
				builder.append(symbol_k);
			else
				builder.append(symbol_q);
			builder.append(':');
			for (int i : selectDiceArray)
				builder.append(i).append(',');
			builder.delete(builder.length()-1,builder.length()).append(']');
		}
		return builder.toString();
	}

	@Override
	public String getTrueExpression()
	{
		if(onlyNumber)
			return String.valueOf(result);
		int[] array;
		if(maxOrMinMode!=GET_ALL_DICE_MODE)
			array=selectDiceArray;
		else if(bonusOrPunishMode!=NO_PUNISH_AND_BONUS)
			array=afterDiceArray;
		else
			array=randomDiceArray;
		if(array.length==1)
			return String.valueOf(array[0]);
		else
		{
			var builder=new StringBuilder();
			builder.append('(');
			for (int i : array)
				builder.append(i).append('+');
			builder.delete(builder.length()-1,builder.length()).append(')');
			return builder.toString();
		}
	}

	@Override
	public String getResourceExpression()
	{
		return resource;
	}

	private Integer getInteger(int index)
	{
		if(index>=list.size()||index<0)
			return null;
		var var=list.get(index);
		if(!(var instanceof Integer))
			return null;
		return (Integer) var;
	}
}

File: src/main/java/com/nobot/plugin/dice/expressionAnalyzer/RandomExpression.kt
package com.nobot.plugin.dice.expressionAnalyzer

import com.nobot.plugin.dice.expressionAnalyzer.SpecialSymbol.*
import java.util.*
import kotlin.jvm.Throws
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

/**
 * 这是最小的随机数表达式单位
 */
class RandomExpression @Throws(ExpressionException::class) constructor(private val random: Random, private val resource: String) : SpecialSymbol, Expression
{
	private var diceNum = 1
	private var diceDownLimit = 0
	private var diceUpperLimit = 0
	private var bonusNum = 0
	private var punishNum = 0
	private var bigDiceNum = 0
	private var smallDiceNum = 0
	private var onlyNumber = false

	private var bonusOrPunishNum = 0
	private var maxOrMinNum = 0
	private lateinit var randomDiceArray: IntArray
	private lateinit var extraDiceArray: IntArray
	private lateinit var afterDiceArray: IntArray
	private lateinit var selectDiceArray: IntArray

	private var result = 0
	override fun getResult(): Int=result

	private val list: ArrayList<Any> = ArrayList()

	override fun calculation()
	{
		if (onlyNumber) return
		randomDiceArray = IntArray(diceNum) { random.nextInt(diceUpperLimit - diceDownLimit + 1) + diceDownLimit }
		if (bonusOrPunishNum != 0)
		{
			extraDiceArray = IntArray(bonusOrPunishNum) { random.nextInt(10) }
			afterDiceArray=IntArray(randomDiceArray.size)
			for (i in randomDiceArray.indices)
			{
				val num = randomDiceArray[i]
				var tenDigit = num / 10
				val singleDigit = num % 10
				tenDigit = if (bonusOrPunishNum > 0) min(tenDigit, extraDiceArray.minOf { it })
				else max(tenDigit, extraDiceArray.maxOf { it })
				afterDiceArray[i] = tenDigit * 10 + singleDigit
			}
		}
		if (maxOrMinNum != 0)
		{
			var arr = if (bonusOrPunishNum != 0) afterDiceArray.copyOf() else randomDiceArray.copyOf()
			if (maxOrMinNum < 0) arr.sort() else arr.sortDescending()
			selectDiceArray = arr.sliceArray(0 until abs(maxOrMinNum))
		}
		result = when
		{
			maxOrMinNum != 0 -> selectDiceArray.sum()
			bonusOrPunishNum != 0 -> afterDiceArray.sum()
			else -> randomDiceArray.sum()
		}
	}

	override fun getShowExpression(): String
	{
		if (onlyNumber) return result.toString()
		val builder = StringBuilder()
		if (randomDiceArray.size != 1)
		{
			builder.append('(')
			for (i in randomDiceArray) builder.append(i).append("+")
			builder.delete(builder.length - 1, builder.length).append(')')
		}
		else builder.append(randomDiceArray[0])
		if (bonusOrPunishNum!=0)
		{
			builder.append('{')
			builder.append(if(bonusOrPunishNum>0) symbol_b else symbol_p)
			builder.append(':')
			for (i in extraDiceArray) builder.append(i).append(',')
			builder.deleteCharAt(builder.length - 1).append('}')
		}
		if (maxOrMinNum!=0)
		{
			builder.append('[')
			builder.append(if (maxOrMinNum>0)symbol_k else symbol_q)
			builder.append(':')
			for (i in selectDiceArray) builder.append(i).append(',')
			builder.deleteCharAt(builder.length - 1).append(']')
		}
		return builder.toString()
	}

	override fun getTrueExpression(): String
	{
		if (onlyNumber) return result.toString()
		val array = when
		{
			maxOrMinNum!=0 -> selectDiceArray
			bonusOrPunishNum!=0 -> afterDiceArray
			else -> randomDiceArray
		}
		return if (array.size == 1) array[0].toString()
		else
		{
			val builder = StringBuilder()
			builder.append('(')
			for (i in array) builder.append(i).append('+')
			builder.deleteCharAt(builder.length-1).append(')')
			builder.toString()
		}
	}

	override fun getResourceExpression(): String= resource

	private fun getInteger(index: Int): Int?
	{
		if (index >= list.size || index < 0) return null
		val `var` = list[index]
		return if (`var` !is Int) null else `var`
	}

	init
	{
		if (symbol_d in resource)
		{
			onlyNumber = false
			val chars = resource.toCharArray()
			val sb = StringBuilder()
			//		分离数字和字符
			for (c in chars)
			{
				when (c)
				{
					'0', '1', '2', '3', '4', '5', '6', '7', '8', '9' -> sb.append(c)
					else ->
					{
						if (sb.isNotEmpty())
						{
							list.add(sb.toString().toInt())
							sb.clear()
						}
						list.add(c)
					}
				}
			}
			if (sb.isNotEmpty()) list.add(sb.toString().toInt())
			//		识别并处理控制符
			for (i in list.indices)
			{
				val `object` = list[i]
				if (`object` is Char)
				{
					when (`object`)
					{
						symbol_b -> bonusNum = getInteger(i + 1) ?: 1
						symbol_p -> punishNum = getInteger(i + 1) ?: 1
						symbol_k -> bigDiceNum = getInteger(i + 1) ?: 1
						symbol_q -> smallDiceNum = getInteger(i + 1) ?: 1
						symbol_d ->
						{
							diceNum = getInteger(i - 1) ?: 1
							diceUpperLimit = getInteger(i + 1) ?: 100
							diceDownLimit = 1
						}
						symbol_t ->
						{
							diceDownLimit = getInteger(i - 1) ?: 1
							diceUpperLimit = getInteger(i + 1) ?: 100
						}
						else -> throw ExpressionException(resource, "无法识别的控制符号：$`object`",null,null)
					}
				}
			}
			//		获得正式的奖励或者惩罚数量，两者相互抵消，检测是否是百分骰
			bonusOrPunishNum = bonusNum - punishNum
			if ((diceDownLimit != 1 || diceUpperLimit != 100 || diceNum != 1) && bonusOrPunishNum != 0)
				throw ExpressionException(resource, "非单独一个百分骰不启用b和p控制符号",null,null)
			//		获得正式的取大或者取小值，同时存在报错，检测是否取值超过了总值
			if (bigDiceNum != 0 && smallDiceNum != 0) throw ExpressionException(resource, "同时存在k和q控制符",null,null)
			when
			{
				bigDiceNum > 0 -> maxOrMinNum = bigDiceNum
				smallDiceNum > 0 -> maxOrMinNum = 0-smallDiceNum
			}
			if (abs(maxOrMinNum) > diceNum) throw ExpressionException(
				resource,
				"取最" + (if (maxOrMinNum > 0) '大' else '小') + "骰子数目超出总数目"
				,null,null)
		}
		else
		{
			onlyNumber = true
			result = try
			{
				resource.toInt()
			}
			catch (e: NumberFormatException)
			{
				throw ExpressionException(resource, "不存在D时无法使用其他修饰符",null,null)
			}
		}
	}
}